### PR TITLE
Update TinkerPop 3.3.1

### LINKED ISSUE #845: Update 0.3 (master) to TinkerPop 3.3.1
- [x] Release TinkerPop 3.3.1
- [x] Pass TinkerPop tests - `mvn clean install -Dtest.skip.tp=false`
- [ ] Update gremlin-server.sh
- [ ] Update data files
- [ ] Update docs

### PR DESCRIPTION
**Pull request by robertdale**
Fixes #845

Signed-off-by: Robert Dale <robdale@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java [hunk 1]
```diff
@@ -98,8 +99,6 @@ public void run() {
         final List<SliceQuery> queries;
         final int numQueries;
         try {
-            job.workerIterationStart(jobConfiguration, graphConfiguration, metrics);
```
Comment (dpitera) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java]: The changes in this class surely have nothing to do with upgrading to TP 3.3.1 right?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/HasStepFolder.java [hunk 2]
```diff
@@ -102,7 +102,6 @@ static void foldInIds(final HasStepFolder janusgraphStep, final Traversal.Admin<
         while (true) {
             if (currentStep instanceof HasContainerHolder) {
                 final HasContainerHolder hasContainerHolder = (HasContainerHolder) currentStep;
-                final Set<Object> ids = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (robertdale) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/HasStepFolder.java]: @dpitera  This is effectively no-op code.  See also https://github.com/JanusGraph/janusgraph/pull/881#pullrequestreview-88333712

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java (hunk 1 above)
Comment (robertdale) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java]: @dpitera  3.2.7/3.3.1 added some OLAP profile() fixes which break some tests. Still working through them actually.

Comment (sjudeng) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java]: Does that mean TP tests are not currently all passing with these updates?

Comment (robertdale) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java]: @sjudeng That is correct.

**Comment by robertdale** [comment]
Maybe someone has more time to progress this faster.  I'm a little stuck on the OLAP stuff and I'm no expert on it.  TinkerPop 3.3.1 introduced a change to the way `profile()` works - https://github.com/apache/tinkerpop/pull/734
The effect is that this TinkerPop test fails: `org.apache.tinkerpop.gremlin.process.traversal.step.map.ProfileTest$Traversals`

The stacktrace is:
```
java.lang.IllegalArgumentException: The memory does not have a value for provided key: 4.1.0(4.0.0())
	at org.apache.tinkerpop.gremlin.process.computer.Memory$Exceptions.memoryDoesNotExist(Memory.java:153)
	at org.janusgraph.graphdb.olap.computer.FulgoraMemory.get(FulgoraMemory.java:130)
	at org.apache.tinkerpop.gremlin.process.computer.traversal.MemoryTraversalSideEffects.add(MemoryTraversalSideEffects.java:84)
	at org.apache.tinkerpop.gremlin.process.computer.traversal.TraversalVertexProgram.workerIterationEnd(TraversalVertexProgram.java:379)
	at org.janusgraph.graphdb.olap.computer.VertexProgramScanJob.workerIterationEnd(VertexProgramScanJob.java:85)
	at org.janusgraph.graphdb.olap.computer.VertexProgramScanJob$Executor.workerIterationEnd(VertexProgramScanJob.java:191)
	at org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor.run(StandardScannerExecutor.java:186)
	at java.lang.Thread.run(Thread.java:748)
```
Actually, the memory does exist. However, FulgoraMemory has that metric with isBroadcast=false thus it can't be 'seen'.  I don't see how to (or if it should) force isBroadcast=true.  The forked (cloned) VertexPrograms don't seem to have this problem.  So it makes me wonder if VertexPrograms can/should have nested, i.e. workerIterationStart, child VertexPrograms (cloned). But because these are also ScanJobs and they (e.g. IndexRemoveJob) rely on the workerIterationStart to initialize some state, it can't be simply un-nested.

If someone wants to pick up on this or just look at it, feel free.  Otherwise I'll plod along as time allows.  Whichever the case, the same fix will need to be applied to #881. I can handle the backport.

**Comment by robertdale** [comment]
And to help anyone getting start on it, here's a shortcut to building and testing:
```
export GREMLIN_TESTS='org.apache.tinkerpop.gremlin.process.traversal.step.map.ProfileTest$Traversals' 
mvn clean install -Dtest.skip.tp=false -DskipTests=true  -Dmaven.javadoc.skip=true
```

**Comment by sjudeng** [comment]
It looks like it can be resolved by ignoring the exception (see https://github.com/sjudeng/janusgraph/commit/53df417a32f6aa1eeda5864628cbcc05c1faddaf). I'm sure there's a better solution but I looked at other options around setting/getting the relevant item in memory and nothing was obvious to me. I tested the full ProcessComputerSuite and, with the exception of two tests in the WhereTest suite, the profile and all other tests pass with the update to ignore the exception. The WhereTest failures don't seem to be related to the above "fix".

```
org.apache.tinkerpop.gremlin.process.traversal.step.filter.WhereTest#g_V_asXaX_outXcreatedX_whereXasXaX_name_isXjoshXX_inXcreatedX_name
org.apache.tinkerpop.gremlin.process.traversal.step.filter.WhereTest#g_V_hasXageX_asXaX_out_in_hasXageX_asXbX_selectXa_bX_whereXa_outXknowsX_bX
```

In both cases the result size is zero instead of the (non-zero) expected size. Not sure what's going on with those. I'll debug but would likewise welcome any insights you might have, like if there was a relevant change in TinkerPop 3.3.1.

**Comment by robertdale** [comment]
@sjudeng Maybe you don't have the other changes in this PR?  All tests pass for me with the catch.  However, I'm not convinced that's the right approach or even correct.

**Comment by sjudeng** [comment]
Yeah I'm a little concerned about it being in the core `StandardScannerExecutor`. Another solution would be to look at moving the catch to FulgoraMemory if possible. Maybe it's flawed but I've always treated FulgoraGraphComputer as purely for testing and as such have not been as concerned with updates (read "possible hacks") necessary to make the (many abundant) tests pass.

I've been interested in starting a discussion to get opinions on whether we could move olap out to new janusgraph-spark or existing janusgraph-hadoop and then move off FulgoraGraphComputer in favor of SparkGraphComputer (we'd have to add a spark import to janusgraph-core to keep it there). I'd say almost all my time and debugging across all the TinkerPop upgrades in Titan/JanusGraph have involved FulgoraGraphComputer and I've always wondered whether it was worth it since (I don't think) the custom computer is used outside of tests.

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java [hunk 3]
```diff
@@ -183,7 +183,12 @@ public void run() {
             if (!Threads.waitForCompletion(processors,TIMEOUT_MS)) log.error("Processor did not terminate in time");
 
             cleanup();
-            job.workerIterationEnd(metrics);
+            try {
+                job.workerIterationEnd(metrics);
+            } catch (Exception e) {
```
Comment (sjudeng) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java]: Change to `IllegalArgumentException`

**Comment by robertdale** [comment]
All tests pass `mvn clean install -Dtest.skip.tp=false -DskipTests=false -pl janusgraph-test -am`

**Comment by sjudeng** [comment]
I usually test across all backends though admittedly I can't remember if there's ever been an issue in the storage backends that wasn't exhibited when using memory (e.g. janusgraph-test).

**Comment by robertdale** [comment]
Yeah, I wouldn't think there would be a problem.  I'll let a test run and update when it's done.

**Comment by robertdale** [comment]
Well, everything passed as far as hbase. Stopping now.  Travelling rest of the day...

```
[INFO] JanusGraph: Distributed Graph Database
[INFO] JanusGraph-Core: Core Library for JanusGraph
[INFO] JanusGraph-Test: Test Suite for JanusGraph
[INFO] JanusGraph CodePipelines CI: Distributed release testing.
[INFO] janusgraph-server
[INFO] JanusGraph-BerkeleyJE: Distributed Graph Database
[INFO] JanusGraph-Cassandra: Distributed Graph Database
[INFO] JanusGraph-CQL: Distributed Graph Database
[INFO] JanusGraph-ElasticSearch: Distributed Indexing Support
[INFO] JanusGraph-HBase: Parent Module
[INFO] JanusGraph-HBase: Version-independent Core
[INFO] JanusGraph-HBase: 0.98 Compatibility Shim
```

**Comment by robertdale** [comment]
@sjudeng Not sure I'm following. You want to merge 0.2 with TP 3.2.7 changes into master then apply TP 3.3.1-specific changes?

**Comment by sjudeng** [comment]
Right, just to make the changeset here more specific and also let you fix the conflicts when you rebase this on master after 0.2 is merged in. I was just going to submit the 0.2->master PR but there are already other conflicts and I can't resolve at the moment. I can do that later unless you or someone else wants to.

**Comment by robertdale** [comment]
Ok, that's what I thought. Just wanted to make sure I was on the same page.  I'll get it.

**Comment by robertdale** [comment]
Pushed as merge. Tests pass.
```
INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] JanusGraph: Distributed Graph Database ............. SUCCESS [  0.328 s]
[INFO] JanusGraph-Core: Core Library for JanusGraph ....... SUCCESS [ 12.610 s]
[INFO] JanusGraph-Test: Test Suite for JanusGraph ......... SUCCESS [30:24 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

**Pull request merged**